flashlight/optim/Optimizers.cpp

```diff
@@ -168,8 +175,24 @@ void AdamOptimizer::step() {
     double correctedBias2 = 1 - std::pow(beta2_, count_);
     double correctedLr = lr_ * std::sqrt(correctedBias2) / correctedBias1;
 
-    data = data - (correctedLr * biasedFirst) / (af::sqrt(biasedSecond) + eps_);
-
+    if (data.type() == f16) {
+      af::array& dataF32 = parametersF32_[i];
+      if (wd_ != 0) {
+        // Weight decay term
+        dataF32 = dataF32 - wd_ * dataF32;
+      }
+      dataF32 = dataF32 -
+          (correctedLr * biasedFirst) / (af::sqrt(biasedSecond) + eps_);
+      af::eval(dataF32);
+      data = dataF32.as(f16);
+    } else {
+      if (wd_ != 0) {
+        // Weight decay term
+        data = data - wd_ * data;
+      }
+      data =
+          data - (correctedLr * biasedFirst) / (af::sqrt(biasedSecond) + eps_);
+    }
```

Comment:
We want to do all parameter and gradient accumulations in fp32 for now. There's a fair amount of research supporting that this is important in seq2seq domains (happy to share); there shouldn't be a need to check if the accumulator is fp16.

Reply:
@jacobkahn sure. Can you please share some of those articles with me to take a look at them?

Reply:
https://arxiv.org/pdf/1710.03740.pdf is the canonical one, but there's also https://arxiv.org/pdf/1802.00930.pdf and https://arxiv.org/pdf/1807.11205.pdf which discuss the update a bit.